impl_core/src/main/java/io/opencensus/implcore/metrics/DoubleGaugeMetricImpl.java

```diff
@@ -0,0 +1,164 @@
+      new ConcurrentHashMap<List<LabelValue>, Point>();
+  private final List<LabelKey> labelKeys;
+
+  DoubleGaugeMetricImpl(String name, String description, String unit, List<LabelKey> labelKeys) {
+    this.metricDescriptor =
+        MetricDescriptor.create(name, description, unit, Type.GAUGE_DOUBLE, labelKeys);
+    this.labelKeys = labelKeys;
+  }
+
+  @Override
+  public Point addPoint(List<LabelValue> labelValues) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  @Override
+  public <T> void addPoint(
+      List<LabelValue> labelValues, @Nullable T obj, ToDoubleFunction<T> function) {
+    checkNotNull(labelValues, "labelValues should not be null.");
+    checkArgument(labelKeys.size() == labelValues.size(), "Incorrect number of labels.");
+
+    addAndGetPoint(labelValues, obj, checkNotNull(function, "function"));
+  }
+
+  @Override
+  public Point getDefaultPoint() {
+    List<LabelValue> labelValues = new ArrayList<LabelValue>(labelKeys.size());
+
+    return addAndGetPoint(labelValues, null, null);
+  }
+
+  private <T> Point addAndGetPoint(
+      List<LabelValue> labelValues, @Nullable T obj, @Nullable ToDoubleFunction<T> function) {
+    Point currentPoint = pointConcurrentHashMap.get(labelValues);
+    if (currentPoint != null) {
+      return currentPoint;
+    } else {
+      Point newPoint =
+          function != null
+              ? new PointImpl<T>(labelValues, obj, function)
+              : new PointImpl<T>(labelValues);
+
+      pointConcurrentHashMap.put(labelValues, newPoint);
+      return newPoint;
+    }
+  }
+
+  private List<TimeSeries> getTimeSeries(Clock clock) {
+    List<TimeSeries> timeSeriesList = new ArrayList<TimeSeries>(pointConcurrentHashMap.size());
+
+    for (Map.Entry<List<LabelValue>, Point> pointEntry : pointConcurrentHashMap.entrySet()) {
+      timeSeriesList.add(
+          TimeSeries.create(
+              pointEntry.getKey(),
+              Collections.singletonList(
+                  io.opencensus.metrics.export.Point.create(
+                      Value.doubleValue(((PointImpl) pointEntry.getValue()).get()), clock.now())),
+              null));
+    }
+    return timeSeriesList;
+  }
+
+  @Override
+  public Metric getMetric(Clock clock) {
+    return Metric.create(metricDescriptor, getTimeSeries(clock));
+  }
+
+  /** Implementation of {@link io.opencensus.metrics.DoubleGaugeMetric.Point}. */
+  public static final class PointImpl<T> extends Point {
+    private final AtomicDouble value = new AtomicDouble(0);
+    private final List<LabelValue> labelValues;
+    @Nullable private T obj;
+    @Nullable private ToDoubleFunction<T> function;
+
+    PointImpl(List<LabelValue> labelValues) {
+      this.labelValues = Collections.unmodifiableList(labelValues);
```

Comment:
PointImpl is private (only trusted APIs call this) so no need to copy the labels here. Actually you should do this before you add the key in the map and use the same instance here.